src/main/java/picard/sam/HitsForInsert.java

```diff
@@ -194,45 +195,100 @@ public void setPrimaryAlignment(final int primaryAlignmentIndex) {
 
-        // Now renumber any correlated alignments, and remove hit index if no correlated read.
-        int hi = 0;
+        renumberHitIndex();
+    }
+
+    /**
+     * Renumber any correlated alignments, and remove hit index if no correlated read.
+     */
+    private void renumberHitIndex() {
         for (int i = 0; i < numHits(); ++i) {
             final SAMRecord first = getFirstOfPair(i);
             final SAMRecord second = getSecondOfPair(i);
-            if (first != null && second != null) {
+            if (nonNull(first) && nonNull(second)) {
                 first.setAttribute(SAMTag.HI.name(), i);
                 second.setAttribute(SAMTag.HI.name(), i);
-                ++hi;
-            } else if (first != null) {
-                first.setAttribute(SAMTag.HI.name(), null);
-            } else {
-                second.setAttribute(SAMTag.HI.name(), null);
+            }
+            if (nonNull(first) && isNull(second)) first.setAttribute(SAMTag.HI.name(), null);
+            if (isNull(first) && nonNull(second)) second.setAttribute(SAMTag.HI.name(), null);
+        }
+    }
+
+    /**
+     * This method lines up alignments in firstOfPairOrFragment and secondOfPair lists so that the first and the second
+     * records of each pair have the same index.
+     */
+    void coordinateByMate() {
+        final List<SAMRecord> newFirstOfPairOrFragment = new ArrayList<>();
+        final List<SAMRecord> newSecondOfPair = new ArrayList<>();
+
+        for (int i = 0; i < firstOfPairOrFragment.size(); i++) {
+            final SAMRecord first = firstOfPairOrFragment.get(i);
+            newFirstOfPairOrFragment.add(i, first);
+            newSecondOfPair.add(null);
+
+            for (int k = 0; k < secondOfPair.size(); k++) {
+                final SAMRecord second = secondOfPair.get(k);
+                if (arePair(first, second)) {
+                    newSecondOfPair.set(i, second);
+                    secondOfPair.set(k, null);
+                    break;
+                }
+            }
+        }
+
+        for (int i = 0; i < secondOfPair.size(); i++) {
+            if (nonNull(secondOfPair.get(i))) {
+                newFirstOfPairOrFragment.add(i, null);
+                newSecondOfPair.add(i, secondOfPair.get(i));
             }
         }
+
+        firstOfPairOrFragment.clear();
+        firstOfPairOrFragment.addAll(newFirstOfPairOrFragment);
+        secondOfPair.clear();
+        secondOfPair.addAll(newSecondOfPair);
+
+        renumberHitIndex();
+    }
+
+    /**
+     * @return true if records belong to the same pairwise alignment
+     */
+    private static boolean arePair(SAMRecord first, SAMRecord second) {
+        return nonNull(first) && nonNull(second)
+                && first.getMateAlignmentStart() == second.getAlignmentStart()
+                && first.getAlignmentStart() == second.getMateAlignmentStart()
+                && Objects.equals(first.getReferenceName(), second.getMateReferenceName())
```

Comment:
Why use `Object.equals()` here? is it possible that one of these is `null`?

Reply:
Since both `SamRecod#mReferenceName` and `SamRecord#mMateReferenceName` [are nullable](https://github.com/samtools/htsjdk/blob/934fb62e8b95071787bad57ee49244adcb655e6b/src/main/java/htsjdk/samtools/SAMRecord.java#L571) we need to ensure that they are non-null or simply delegate it to `Objects.equals` to avoid `NullPointerException`. This is a theoretical case that may hardly ever occur in runtime. So, do we need to prevent this case or can I unwrap this check?

Reply:
It's OK, I was wondering if this was added to fix a known problem or was just precautionary.

Looking at `SamRecord` I see that it's common to check these fields for `null` so it seems wise to do it here too.

It would be great if picard had nullable constraints in the code base! That would be a huge PR though.

Reply:
`Optional` can be rather helpful here and it would be just great if we could use [`Try`](https://github.com/vavr-io/vavr/blob/master/vavr/src/main/java/io/vavr/control/Try.java) and [`Lazy` ](https://github.com/vavr-io/vavr/blob/master/vavr/src/main/java/io/vavr/Lazy.java) types from [vavr](https://github.com/vavr-io/vavr) to reduce service-code and to clarify Picard logic!=D